variant/effect.go

```diff
@@ -0,0 +1,276 @@
+func (i Insertion) Effect(codingSeq []dna.Base, offsetStart int, offsetEnd int) (CodingChange, error) {
+	var answer CodingChange
+	offsetPos := i.Pos + offsetStart
+	answer.CodingPos = offsetPos
+	answer.ProteinPos = offsetPos / 3
+	var protOffset int
+
+	if offsetPos < 0 {
+		return answer, ErrNegPos
+	}
+
+	if offsetPos > len(codingSeq) {
+		return answer, ErrInvalidPosition
+	}
+
+	frame := offsetPos % 3
+	codonStart := offsetPos - frame
+
+	switch {
+	case len(i.Seq)%3 != 0: // frameshift
+		answer.Type = Frameshift
+		shiftedSeq := make([]dna.Base, len(i.Seq)+len(codingSeq[codonStart:]))
+		copy(shiftedSeq[:frame], codingSeq[codonStart:offsetPos])  // add bases before insertion
+		copy(shiftedSeq[frame:frame+len(i.Seq)], i.Seq)            // add inserted bases
+		copy(shiftedSeq[frame+len(i.Seq):], codingSeq[offsetPos:]) // add bases after insertion
+		answer.RemovedAa, answer.AddedAa, protOffset = aaChange(codingSeq[codonStart:], shiftedSeq)
+
+	case frame != 0: // disrupts a codon
+		answer.Type = InFrameInsertion
+		insSeq := make([]dna.Base, len(i.Seq)+3)
+		copy(insSeq[:frame], codingSeq[codonStart:offsetPos])              // add bases before insertion
+		copy(insSeq[frame:frame+len(i.Seq)], i.Seq)                        // add inserted sequence
+		copy(insSeq[frame+len(i.Seq):], codingSeq[offsetPos:codonStart+3]) // add bases after insertion
+		answer.RemovedAa, answer.AddedAa, protOffset = aaChange(codingSeq[codonStart:codonStart+3], insSeq)
+
+	default: // in frame & does not disrupt a codon
+		answer.Type = InFrameInsertion
+		answer.RemovedAa, answer.AddedAa, protOffset = aaChange(nil, i.Seq)
+	}
+
+	answer.ProteinPos += protOffset
+	return answer, nil
+}
+
+// Effect for Deletion determines a coding change resulting from a deletion.
+func (d Deletion) Effect(codingSeq []dna.Base, offsetStart int, offsetEnd int) (CodingChange, error) {
+	var answer CodingChange
+	offsetStartPos := d.Start + offsetStart
+	offsetEndPos := d.End + offsetEnd
+	answer.CodingPos = offsetStartPos
+	answer.ProteinPos = offsetStartPos / 3
+	var protOffset int
+
+	if offsetStartPos < 0 {
+		return answer, ErrNegPos
+	}
+
+	if offsetEndPos > len(codingSeq) {
+		return answer, ErrInvalidPosition
+	}
+
+	delLen := offsetEndPos - offsetStartPos
+	startFrame := offsetStartPos % 3
+	endFrame := (offsetEndPos - 1) % 3
+	codonStart := offsetStartPos - startFrame       // start of first codon affected (closed)
+	codonEnd := ((offsetEndPos - 1) - endFrame) + 3 // end of last codon affected (open)
+
+	switch {
+	case delLen%3 != 0: // frameshift
+		answer.Type = Frameshift
+		shiftedSeq := make([]dna.Base, len(codingSeq[codonStart:])-delLen)
+		copy(shiftedSeq[:startFrame], codingSeq[codonStart:offsetStartPos]) // add bases before deletion
+		copy(shiftedSeq[startFrame:], codingSeq[offsetEndPos:])             // add bases after deletion
+		answer.RemovedAa, answer.AddedAa, protOffset = aaChange(codingSeq[codonStart:], shiftedSeq)
```

Comment:
I think aaChange handles the case of the stop codon being deleted, but I didn't see code handling the start codon being deleted.  From my quick look, it seemed like we would still start translating, even though there is no start codon

Reply:
Yes, I was not quite sure how to handle this. There are cases of non-AUG start codons so I felt it was safer to let this be handled downstream by checking if ProteinPos == 0. Open to alternative ideas however.